### PR TITLE
Check for consent in onRouteUpdate

### PR DESCRIPTION
**Pull request by pm990320**
Currently the plugin pushes the consent given event on every client-side page navigation, irrespective of whether or not the user has given consent. This is incompatible with the behaviour of the `consentOnContinuedBrowsing: false` option. This PR resolves this issue by checking using the API if consent has been given before triggering.

I've provided some integration tests, please let me know if I should duplicate this in the `custom-gtm-test` project.

Thanks!

### CODE REVIEW

File: test-projects/gtm-test/src/pages/index.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ const IndexPage = () => (
     <div style={{ maxWidth: `300px`, marginBottom: `1.45rem` }}>
       <Image />
     </div>
-    <Link to="/page-2/">Go to page 2</Link>
+    <Link to="/page-2/" className="page-2-link">Go to page 2</Link>
```
Comment (NoriSte) [inline test-projects/gtm-test/src/pages/index.js:17]: thought: Removing the class.

In general, as much as we can, we should leverage the same things the user would do. The user would click on the "Go to page 2" link, so we should retrieve it with `cy.contains('Go to page 2')`.

File: test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    cy.window().then(win => {
+      win.dataLayer = [];
+    });
+
+    cy.get(".page-2-link").click();
+
+    cy.window().its('dataLayer').should('contain', { event: "iubenda_consent_given" });
+  })
+
+  it('The Iubenda cookie footer should NOT push the event to the GTM data layer if consent has NOT been given and the user navigates to a different page (onRouteUpdate)', () => {
```
Comment (NoriSte) [inline test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js:28]: thought: transform "NOT" to "not"

There is no need to have "NOT" capitalized

**Comment by pm990320** [comment]
Thanks @NoriSte - agree with all of those. I've made those changes, please have a look!

File: test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js [hunk 3]
```diff
@@ -20,22 +19,21 @@ context('Iubenda GTM client-side navigation', () => {
...
-    cy.get(".page-2-link").should("exist");
 
     cy.window().then(win => {
       win.dataLayer = [];
     });
 
-    cy.get(".page-2-link").click();
+    cy.contains('Go to page 2').click();
 
     cy.window().its('dataLayer').should('not.contain', { event: "iubenda_consent_given" });
```
Comment (NoriSte) [inline test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js:38]: I thought about that this night, tell me what do you think: generally speaking, this check is weak, what happens if the plugin pushes something in the array a millisecond after the check? The problem is not with your test, obviously, it's in the nature of testing something that does not happen 😉

How could we fix it? My idea is:
- the first test (the "positive" test, not the "negative" test we are talking about) should check that
  - if you visit the first page and you give the consent, the array contains the expected object
  - then, if you navigate to the second page, the array contains two times the expected object
- while the "negative" test should check that
  - if you visit the first page and you don't give the consent...
  - then, if you visit the second page and don't you give the consent...
  - then, if you visit the first page (or a third one, it doesn't matter) and you give the consent, the array contains the expected object
  - then, if you navigate to the second page (or a fourth one, it doesn't matter), the array contains two times the expected object

This way we would be completely sure that the first and the second navigation don't pollute the array! What do you think?

Comment (pm990320) [reply test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js:38]: You're right, there is a potential for a race condition here.

I agree that we can check for duplication on the positive test, however, I don't think the "negative" test should involve giving consent at any point - it seems to defeat the purpose of the negative test. By adding the third visit, I don't think it fixes the issue of the race condition either.

Perhaps we could add a third/fourth visit **without** giving consent, to give the event handler multiple chances to fire? We could also add a `cy.wait(1000)` but I would rather avoid this approach as it doesn't really solve the problem either.

What do you think?

Comment (NoriSte) [reply test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js:38]: You're right. To have the `cy.wait` advantages without leaving the Cypress garden, the second test could:
- visit the first page
- go to the second page
- the test expects to fail ([here](https://github.com/NoriSte/cypress-wait-until/blob/master/cypress/integration/plugin.spec.js#L202) you can find a real world example)
```js
 cy.once('fail', (err) => {
    expect(err.message).to.be.equal('<PUT THE CYPRESS ERROR HERE>')
  })
```
- assert that the array should contain the expected object
- Cypress retries up to 5 seconds for the array check, then it fails
- thanks to `cy.once('fail' // ...` our test will pass

Doing this: we mitigate the race condition (if the array is empty even after 5 seconds we could be enough sure that Iubenda won't push the object at all) and we stay in what Cypress offers. Do you agree or do you prefer going with your solution? 😊

Comment (pm990320) [reply test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js:38]: I like this! Will give that a try. Thanks

**Comment by NoriSte** [comment]
Hi Patrick! Thanks a lot! I used your PR to fresh up the whole plugin, updating all the dependencies, and fixing the tests. I run the Travis build with #329 just to be sure that everything works and I made some minor changes to your tests too. Could you give me writing access to your fork in order to allow me to push the commits on your branch, please? So we can continue discussing about it here and once we agree we could squash the commits 😊

**Comment by pm990320** [comment]
No problem, you should have access now!

**Comment by NoriSte** [comment]
The relevant changes are https://github.com/NoriSte/gatsby-plugin-iubenda-cookie-footer/pull/326/commits/4c68fd84e9659bbb5563f834ef82ae56ad19e5b1 and https://github.com/NoriSte/gatsby-plugin-iubenda-cookie-footer/pull/326/commits/3d49599066c2366606efcfe8be85a25ff6b8bf5a where I change the way you assert about the event and I aggregate the tests. I changed the assertions because I had a lot of problems with the previous assertions (not only yours, with mine too) and I've aggregated the tests to keep them denser. You could argue with the latter 😉

**Comment by pm990320** [comment]
Ok, yes this looks like a better way of doing it, given the dataLayer would not be defined! I'm happy with this. Thanks for doing the cleanup

**Comment by NoriSte** [comment]
@all-contributors please add @pm990320 for bug, code, review, test

**Comment by NoriSte** [comment]
I'll wait for the Travis build then I'll add you to the contributors and I can release it 😊 Thank you so much @pm990320 👏👏👏

**Comment by NoriSte** [comment]
@all-contributors please add @NoriSte for ideas, code, test, doc

**Comment by NoriSte** [comment]
@all-contributors please add @pm990320 for bug, code, review, test

**Comment by NoriSte** [comment]
@pm990320 completed and published, finally ❤️

**Pull request merged**